src/dust/coagulation/coagulation.cpp

```diff
@@ -0,0 +1,308 @@
+  cpars.vfrag = pin->GetOrAddReal("dust", "vfrag", 1.e3); // cm/s
+  cpars.nlim = 1e10;
+  cpars.integrator = pin->GetOrAddInteger("dust", "coag_int", 3);
+  cpars.use_adaptive = pin->GetOrAddBoolean("dust", "coag_use_adaptiveStep", true);
+  cpars.mom_coag = pin->GetOrAddBoolean("dust", "coag_mom_preserve", true);
+  cpars.nCall_mx = pin->GetOrAddInteger("dust", "coag_nsteps_mx", 1000);
+  // dust particle internal density g/cc
+  const Real rho_p = pin->GetOrAddReal("dust", "grain_density", 1.25);
+  cpars.rho_p = rho_p;
+
+  cpars.ibounce = pin->GetOrAddBoolean("dust", "coag_bounce", false);
+
+  int coord_type = 0; // density
+
+  const bool isurface_den = pin->GetOrAddBoolean("dust", "surface_density_flag", true);
+  if (isurface_den) coord_type = 1;
+  cpars.coord = coord_type; // 1--surface density, 0: 3D
+
+  cpars.err_eps = 1.0e-1;
+  cpars.S = 0.9;
+  cpars.cfl = 1.0e-1;
+  cpars.chi = 1.0;
+
+  // some checks and definition
+  if (cpars.use_adaptive) {
+    if (cpars.integrator == 3) {
+      cpars.pgrow = -0.5;
+      cpars.pshrink = -1.0;
+    } else if (cpars.integrator == 5) {
+      cpars.pgrow = -0.2;
+      cpars.pshrink = -0.25;
+    } else {
+      std::stringstream msg;
+      msg << "### FATAL ERROR in dust coagulation initialization: " << std::endl
+          << "###   You can not use this integrator with adaptive step sizing: "
+          << cpars.integrator << std::endl;
+      PARTHENON_FAIL(msg);
+    }
+    cpars.errcon = std::pow((5. / cpars.S), (1. / cpars.pgrow));
+  }
+
+  const Real s_min = pin->GetReal("dust", "min_size");
+  const Real s_max = pin->GetReal("dust", "max_size");
+  const Real mmin = 4.0 * M_PI / 3.0 * rho_p * std::pow(s_min, 3);
+  const Real mmax = 4.0 * M_PI / 3.0 * rho_p * std::pow(s_max, 3);
+  const Real cond = 1.0 / (1.0 - nm) * std::log(mmin / mmax);
+  const Real conc = std::log(mmin);
+  if (std::exp(cond) > std::sqrt(2.0)) {
+    std::stringstream msg;
+    msg << "### FATAL ERROR in dust with coagulation: using nspecies >"
+        << std::log(mmax / mmin) / (std::log(std::sqrt(2.0))) + 1. << " instead of " << nm
+        << std::endl;
+    PARTHENON_FAIL(msg);
+  }
+
+  ParArray1D<Real> dust_size("dustSize", nm);
+  auto dust_size_host = dust_size.GetHostMirror();
+
+  for (int n = 0; n < nm; ++n) {
+    Real mgrid = std::exp(conc + cond * n);
+    dust_size_host(n) = std::pow(3.0 * mgrid / (4.0 * M_PI * cpars.rho_p), 1. / 3.);
+  }
+  dust_size.DeepCopy(dust_size_host);
+
+  // allocate array and assign values
+  cpars.klf = ParArray2D<int>("klf", nm, nm);
+  cpars.mass_grid = ParArray1D<Real>("mass_grid", nm);
+  const int n2DRv = coag2DRv::last2;
+  cpars.coagR3D = ParArray3D<Real>("coagReal3D", n2DRv, nm, nm);
+  cpars.cpod_notzero = ParArray3D<int>("idx_nzcpod", nm, nm, 4);
+  cpars.cpod_short = ParArray3D<Real>("nzcpod", nm, nm, 4);
+
+  if (Dust::cgsunit == NULL) {
+    Dust::cgsunit = new CGSUnit;
+  }
+  if (!Dust::cgsunit->isSet()) {
+    Dust::cgsunit->SetCGSUnit(pin);
+  }
```

Comment:
I think we should hold off on doing unit conversions / setting units until we have an artemis units package. Until then, the user will have to set everything to cgs.

Reply:
The dust coagulation uses CGS unit. The user input is "au" and "msun".  I can assume the artemis using CGS unit and let user modify their input if you prefer that.

Reply:
Assume for now that the user will put everything in in cgs. When we add the units capability, we can do the conversions then.

Reply:
Yeah lets ignore any details about units here and just expect a specific unit from the user input file. These dust-specific unit conversions will be replaced by the global #24.

Reply:
Answers my question above.